cpp/inc/bond/ext/grpc/detail/service_call_data.h

```diff
@@ -81,30 +81,32 @@ struct service_unary_call_data : io_manager_tag
     {
         if (ok)
         {
-            // capture the data associated with this one incomming request
-            // so that we can pass it to the user callback. The unary_call
-            // that we create to pass to the user callback takes ownership
-            // of this data.
-            //
-            // This scope block is here so that the shared_ptr that we
-            // create to capture in the lambda is destroyed after we enqueue
-            // in the thread pool but before we schedule the next receive.
-            // When we can use C++14, we can simplify this by using lambda's
-            // capture-by-move.
+            // Capture the data associated with this one incomming request
+            // so that we can pass it to the user callback. When we create
+            // the unary_call to pass to the user callback, the unary_call
+            // and unary_call_impl start collaborating to manage the
+            // lifetime of the unary_call_impl.
             {
-                std::shared_ptr<unary_call_impl<TRequest, TResponse>> receivedCall = _receivedCall;
+                // We have to manually release the pointer from
+                // _receivedCall to capture it in the lambda. When we can
+                // use C++14, we can simplify this by using lambda's
+                // capture-by-move.
+                auto receivedCall = _receivedCall.release();
```

Comment:
I would avoid doing this, because it will require all the scheduled functions to be executed in order to avoid memory leak, or ever worse functional misbehavior. Given that `unary_call_impl` already has a reference counter you can wrap it into `boost::intrusive_ptr` (which is copyable) and capture it that way. This will make sure that the lifetime of that object does not rely on the function invocation. Also note that even with C++14 support you might not be able to move-capture if the scheduler uses `std::function` to store the functions which is unable to store move-only objects.

Reply:
> it will require all the scheduled functions to be executed in order to avoid memory leak

At the moment, it's [already a requirement that the scheduler invoke all callbacks scheduled to avoid memory leaks](https://github.com/Microsoft/bond/blob/70ced821b31a38a93582f5885d7b22ca9ab7638c/cpp/inc/bond/ext/grpc/thread_pool.h#L35-L38). Should this be revisited? Yes. In this PR, I'd rather not: we're not making anything any worse, and we are making the unary_call interface better. Not fixing this now also doesn't prevent or make it harder to fix it later.

> Given that `unary_call_impl` already has a reference counter you can wrap it into `boost::intrusive_ptr` (which is copyable) and capture it that way.

Let me see if I can rework this. I think I need to adjust the refcounting to make this work, but it will end up even better than before.